### PR TITLE
TS-4042: Add feature to buffer request body before making downstream requests

### PR DESCRIPTION
**Pull request by bgaff**
We need a way to examine the request body without making a downstream request, this feature has many use cases including:
- Ability to buffer the body and ensure a full post is received before committing downstream resources.
- Ability to choose an origin based on request body
- Ability to do request content filtering such as a WAF might provide before the origin is involved.

Today you have two options to inspect a request body:

  1) Transformations: the problem with transformations is that you only start receiving the request bytes after a sink has been established, which in this case is the downstream origin.
  2) Create an intercept and use fetch apis to then send the downstream request: while this technically works it turns out to be a ton of code and is in general pretty problematic, we actually tried this approach for a while and had nothing but problems with it.

We feel it would be ideal if we could intercept the body without breaking the normal ATS state flow. There used to exist code (and it's still in the core just #ifdefed out) to drain the request body. I use that code as the basis for this request buffering code. We added APIs to both the C and C++ APIs so that this request buffering can be enabled from a plugin and the plugin can inspect the body as chunks arrive or when it's complete. We've included an example plugin that will error a transaction if a minimum rate of transfer is not maintained. We've been using a very similar method in the core for buffer request bodies for several months without issues so the code that is new (for us) is basically all the API stuff.

I'm confident that this feature will bring plenty of questions / feedback, so let's get that party started. @zwoop @SolidWallOfCode @jpeach @sudheerv : if you have time would you mind commenting / reviewing.

cc. @jacksontj @zizhong @canselcik

### CODE REVIEW

**Comment by jacksontj** [comment]
Glad to see this PR finally make its way out :) We've been running with the base for this PR for quite a while. In addition to simplifying the code-- it really cleans up metrics/debugging as we are no longer required to loop requests back through the state machine.

**Comment by jpeach** [comment]
I quickly scanned the diff. The API changes will still need to go through API review on dev@. Documentation for the new APIs should be done as an API man page in the main documentation rather than in headerdoc (perfectly OK to do that as a subsequent patch).

**Comment by bgaff** [comment]
@jpeach sure thing. Does anyone else have comments regarding the approach before we discuss the individual APIs?

**Comment by sudheerv** [comment]
+1 on the approach.

A specific comment on a quick scan of the code - it seems like the patch only handles the case with *Content-Length* header, what about Chunked-Encoding or H2/SPDY scenarios that don't present the Content-Length header?

**Comment by bgaff** [comment]
@sudheerv , the chunked encoding case is one that I don't think any browser actually does, have you ever seen a browser make a request w/ chunked encoding if so I'll have to look into that case.

**Comment by bgaff** [comment]
@sudheerv, I've asked @zizhong to help in addressing the transfer-encoding chunked case, we should have an update soon.

**Comment by bgaff** [comment]
@sudheerv / @jpeach this pull request has been updated w/ tests for the chunked encoding case that @sudheerv was concerned about. Please let me know if you have any other questions about this?

**Comment by sudheerv** [comment]
@bgaff - Thanks, I'd have to double check though that the change made automatically handles a H2/SPDY upload. There's no Content-Length, nor even a TE header in those cases (although, perhaps, ATS implementation may make it appear like CHUNKED_ENCODING from FetchSM to HttpSM layer?)

**Comment by bgaff** [comment]
That's correct since both use fetchsm if just works like a normal http 1.1 request, as that's what fetch sm is. I've verified this functionality.

**Comment by sudheerv** [comment]
Cool, thanks!

**Comment by bgaff** [comment]
Any other questions about this?

**Comment by bryancall** [comment]
Talking at the github meeting.  We should have a maximum size that will be buffered in memory.  This should go through API review.  Please provide documentation on the APIs.

**Comment by zwoop** [comment]
ping on API review?

**Comment by zwoop** [comment]
@bgaff What do you want to do with this? It's been sitting here for quite a while.

File: plugins/request_buffer/request_buffer.cc [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      int64_t data_len = 0;
+      const char *char_data = NULL;
+      TSIOBufferBlock block = TSIOBufferReaderStart(buffer_reader);
+      while (block != NULL) {
+        char_data = TSIOBufferBlockReadStart(block, buffer_reader, &data_len);
+        memcpy(body + consumed, char_data, data_len);
+        consumed += data_len;
+        block = TSIOBufferBlockNext(block);
+      }
+      // play with the body
```
Comment (bryancall) [inline plugins/request_buffer/request_buffer.cc]: Why are you copying the body if you aren't doing anything with it?

File: proxy/InkAPI.cc [hunk 2]
```diff
@@ -8903,3 +8910,46 @@ TSVConnReenable(TSVConn vconn)
     }
   }
 }
+
+tsapi char *
+TSHttpTxnGetClientRequestBody(TSHttpTxn txnp, int *len)
```
Comment (bryancall) [inline proxy/InkAPI.cc]: Why are you adding this API if you are not using it?

**Comment by bryancall** [comment]
Looking over the pull request I didn't see any limits on how much is going to be buffered on the server.  It might be good to start the transfer to the origin once a configurable limite gets reached.

**Comment by jacksontj** [comment]
@bryancall it'd be nice if we could have some way to "re-enable" the bytes-- such that a plugin could either buffer everything or stream (since the "buffer everything" case is a possible use-case.

**Comment by zizhong** [comment]
@bgaff @bryancall @zwoop What are the remaining works to get this pull request merged? Is there anything I can help with?

As a summary in this thread, 

1. add a `limit` for the buffered body;
2. API review.

What else?

**Comment by bryancall** [comment]
I haven't seen an update to this PR for awhile.  I will close it in a week if it hasn't been updated.

**Comment by bgaff** [comment]
I can help, @Zizhong what of the previous asks haven't been completed?

**Comment by zizhong** [comment]
Thanks, @bgaff . After reading the previous comments, I think @bryancall suggested there should be a `limit` for the request body. And have you guys done `API review` about this PR?

**Comment by zizhong** [comment]
@bryancall @bgaff @zwoop @SolidWallOfCode @jpeach @sudheerv Local patches of `request buffer` we have in Linkedin ATS repo really hurt us a lot(merge issues, etc). I'll work on updating this PR and push it out.
Before that, just want to make sure that upstream still agree on this approach.

**Pull request closed without merging**